### ISSUE TITLE
[FEATURE]Remove double converting from byte[] to float[], and float[] to byte[]  for script scoring with binary vector

### ISSUE THREAD

**Issue by heemin32** [opensearch-project/k-NN#1827]
In https://github.com/opensearch-project/k-NN/pull/1826 which support script scoring on knn binary vector, we convert byte[] to float[] and again float[] to byte[] which will add a latency during query using script scoring on knn binary vector. We want to avoid the unnecessary converting.

**Comment by navneet1v** [comment]
@heemin32 why we are not resolving this as part of 2.16 when this impacts the latency?

**Comment by heemin32** [comment]
It is not a trivial change and 2.16 code freeze is just a next Monday. Score script is already slow compared to ANN and the impact on latency due to double converting could be negligible though actual profiling might be needed to confirm.

**Comment by navneet1v** [comment]
> and the impact on latency due to double converting could be negligible

till we don't have benchmarks we cannot say this. and leaving optimization out of a release doesn't seem correct to me.

**Comment by heemin32** [comment]
> > and the impact on latency due to double converting could be negligible
> 
> till we don't have benchmarks we cannot say this. and leaving optimization out of a release doesn't seem correct to me.

Optimization can be implemented incrementally. That is how it has been till now.

**Comment by VijayanB** [comment]
@kasundra07 is working on this.

**Comment by kasundra07** [comment]
You can assign this to me.

**Comment by kasundra07** [comment]
## Details

[Here](https://github.com/opensearch-project/k-NN/blob/9276c775a325ac516ee4c1edd760a03043c16bce/src/main/java/org/opensearch/knn/index/KNNVectorDVLeafFieldData.java#L64), in `KNNVectorDVLeafFieldData.java,` we create `KNNVectorScriptDocValues` object and depending on the version of opensearch the index was created on, we either use LuceneKNNVectorsFormats - `KNNByteVectorScriptDocValues` for byte/binary vectors & `KNNFloatVectorScriptDocValues` for float vectors (OS version >= 2.17) or `KNNNativeVectorScriptDocValues` (OS version < 2.17).

```
static boolean useLuceneKNNVectorsFormat(final Version indexCreatedVersion) {
    return indexCreatedVersion.onOrAfter(Version.V_2_17_0);
}
```

So, for binary vectors, either `KNNByteVectorScriptDocValues` or `KNNNativeVectorScriptDocValues` will be used depending on the OS version on which the index was created.

For both these classes, we have doGetValue() method where we convert the underlying vector to float[] -

KNNByteVectorScriptDocValues -

```
protected float[] doGetValue() throws IOException {
    byte[] bytes = values.vectorValue();
    float[] value = new float[bytes.length];
    for (int i = 0; i < bytes.length; i++) {
        value[i] = (float) bytes[i];
    }
    return value;
}
```

KNNNativeVectorScriptDocValues -

```
protected float[] doGetValue() throws IOException {
    return getVectorDataType().getVectorFromBytesRef(values.binaryValue());
}

In VectorDataType.java -> BINARY

public float[] getVectorFromBytesRef(BytesRef binaryValue) {
    float[] vector = new float[binaryValue.length];
    int i = 0;
    int j = binaryValue.offset;

    while (i < binaryValue.length) {
        vector[i++] = binaryValue.bytes[j++];
    }
    return vector;
}
```

And for binary vector, script scoring is supported for hamming space type and here we convert the float[] back to byte[] for calculating hamming distance -

```
(float[] q, float[] v) -> 1 / (1 + KNNScoringUtil.calculateHammingBit(toByte(q), toByte(v)))
```

We want to avoid this double conversion, byte[] to float[] & float[] to byte[].

## Solution

Currently `Hamming` class extends abstract class `KNNFieldSpace` which has getScoreScript() method that returns the scoring script for the query -

```
public ScoreScript getScoreScript(
    Map<String, Object> params,
    String field,
    SearchLookup lookup,
    LeafReaderContext ctx,
    IndexSearcher searcher
) throws IOException {
    return new KNNScoreScript.KNNVectorType(params, this.processedQuery, field, this.scoringMethod, lookup, ctx, searcher);
}
```

Here, processedQuery and scoringMethod are bound to float[] vectors -

```
private float[] processedQuery;
private BiFunction<float[], float[], Float> scoringMethod;
```

Since we need byte[] alternatives of these for Hamming, we can instead implement the interface `KNNScoringSpace` and have custom implementation of score script, `KNNByteVectorType`, dealing with byte[] vectors.

And within the execute() method of this new scoring script class (KNNByteVectorType), we can call **getByteValue()** instead of **getValue()** on `KNNVectorScriptDocValues` instance to fetch the byte[] vector and pass it to the scoring method.

So, both the classes mentioned earlier - `KNNByteVectorScriptDocValues` & `KNNNativeVectorScriptDocValues`, would have an additional method **doGetByteValue()** overridden which returns byte[] vector which would be used in the execute() method as explained above.

## Results

Dataset used - <a href="http://corpus-texmex.irisa.fr/">http://corpus-texmex.irisa.fr/</a> (<b>ANN_SIFT1M</b>)<br>Dimensions - 128<br>Vectors - 1,000,000<br><br>Got two indices - one created on OS 2.16 and another created on OS 2.18, and ran the script score query for 10k query vectors with and without the above mentioned changes for both these indices.

<table><thead>
  <tr>
    <th rowspan="2"></th>
    <th colspan="3">With default implementation</th>
    <th colspan="3">With the new changes</th>
    <th rowspan="2">% change in total time</th>
  </tr>
  <tr>
    <th>p50 (in ms)</th>
    <th>p90 (in ms)</th>
    <th>Total (in s)</th>
    <th>p50 (in ms)</th>
    <th>p90 (in ms)</th>
    <th>Total (in s)</th>
  </tr></thead>
<tbody>
  <tr>
    <td>Index_2_16</td>
    <td>190</td>
    <td>196</td>
    <td>1915.408</td>
    <td>73</td>
    <td>74</td>
    <td>736.148</td>
    <td>61.56704</td>
  </tr>
  <tr>
    <td>Index_2_18</td>
    <td>186</td>
    <td>193</td>
    <td>1876.467</td>
    <td>68</td>
    <td>69</td>
    <td>685.511</td>
    <td>63.468</td>
  </tr>
</tbody>
</table>

@VijayanB does this approach seem reasonable to you?

**Comment by jmazanec15** [comment]
@kasundra07 Why not create a byte implementation that mirrors:
```
@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
public abstract class KNNVectorScriptDocValues extends ScriptDocValues<float[]> {
```

Something like 
```
@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
public abstract class KNNByteVectorScriptDocValues extends ScriptDocValues<float[]> {
```

**Comment by kasundra07** [comment]
@jmazanec15 Yes, we can create a byte implementation like you mentioned. I just wanted to reuse the existing class and add one new method to return the byte vector.

**Comment by jmazanec15** [comment]
IMO working with byte[] with interface that defines float[] is going to cause problems. Given that they are separate data types, I think it makes sense to split at that level. We can abstract out/use templates for re-using code.
